verification.py

```diff
@@ -0,0 +1,262 @@
+        self.samples = samples
+        self.sequence_length = 100_000
+        self.recombination_rate = recombination_rate
+        if num_replicates is None:
+            num_replicates = 1
+        self.num_replicates = num_replicates
+
+    def tmrca(self):
+        """
+        Get vector of tmrcas, one for each replicate.
+        """
+        raise NotImplementedError
+
+    def sfs(self):
+        """
+        Get 1D SFS vector.
+        """
+        raise NotImplementedError
+
+    def _tskit_sfs(self, ts_iter):
+        n = sum(self.samples.values())
+        fs = np.zeros(n + 1)
+        for ts in ts_iter:
+            fs += ts.allele_frequency_spectrum(mode="branch", polarised=True)
+        return fs / self.num_replicates
+
+    def _tskit_tmrca(self, ts_iter):
+        tmrca = []
+        for ts in ts_iter:
+            for tree in ts.trees():
+                tmrca.append(tree.time(tree.root))
+        return np.array(tmrca)
+
+    def num_trees(self):
+        return np.array([ts.num_trees for ts in self.ts_list])
+
+
+class SimMs(Simulator):
+    _ms_command = "ms"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # self.N0 = self.graph.demes[0].epochs[0].start_size
+        self.N0 = 1
+        samples = [self.samples.get(deme.name, 0) for deme in self.graph.demes]
+        nsam = sum(samples)
+        assert nsam >= 2
+        r = self.recombination_rate
+        if r is None:
+            # We must set a recombination rate, otherwise the output format
+            # is different and then is not recognised by tsconvert.
+            r = 1e-30
+        rho = 4 * self.N0 * r * (self.sequence_length - 1)
+
+        ms_args = demes.to_ms(self.graph, N0=self.N0, samples=samples)
+        cmd = (
+            f"{self._ms_command} {nsam} {self.num_replicates} {ms_args} "
+            f"-T -r {rho} {self.sequence_length} -p 12"
+        )
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            tmpfile = pathlib.Path(tmpdir) / "ms.out"
+            with open(tmpfile, "w") as f:
+                subprocess.run(
+                    cmd.split(), check=True, stdout=f, stderr=subprocess.DEVNULL
+                )
+            with open(tmpfile) as f:
+                out = f.read()
+        out_reps = re.split(r"^//$", out, flags=re.MULTILINE)[1:]
+        assert len(out_reps) >= 1
+        self.ts_list = [tsconvert.from_ms(rep) for rep in out_reps]
+
+    def sfs(self):
+        return self._tskit_sfs(self.ts_list) * self.N0
+
+    def tmrca(self):
+        return self._tskit_tmrca(self.ts_list) * self.N0
+
+
+class SimMspms(SimMs):
+    _ms_command = "mspms"
+
+
+class SimMsprime(Simulator):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        demog = msprime.Demography.from_demes(self.graph)
+        ts = msprime.sim_ancestry(
+            demography=demog,
+            samples=self.samples,
+            ploidy=1,
```

Comment:
This will make things off by a factor of 2, won't it? Probably simpler to run in diploid mode, but state your samples are haploid using the SampleSet notation: https://tskit.dev/msprime/docs/latest/ancestry.html#sec-ancestry-ploidy

Reply:
Thanks, that seems to do the trick. But I'm very confused now. Shouldn't deme `a` and `b` both be affected? Why does matter which deme is being sampled from?

Reply:
I've no idea, I didn't think about the details, I just scanned for obvious problems :smile:

Reply:
I'm actually a little concerned about this, because it suggests that changing the ploidy between 1/2 can result in scaling of time that *isn't* different by a factor of 2. If you look at the plots above, the stats look fine when sampling from `a`, and the stats for `b` are off by some factor that is clearly not 2.

Reply:
I'll translate this into an msprime-only MWE and open an msprime discussion.